### DISCUSSION TITLE
Alpine Expression Error: prevKeys is undefined

### DISCUSSION THREAD

**Discussion by Ronan-Lenor** [alpinejs/alpine#2628] [1. Help]
Hello,

i'm using Livewire to generate an HTML code that look like so : (basically, datalist & template tags are regenerated on each ajax request)
```
<div wire:loading.remove>
  <datalist id="dropdown">
      @foreach ($this->paginator as $model)
          <option value="{{ $model->id }}">{{  $model->title }}</option>
      @endforeach
  </datalist>
  <template x-for="(option,index) in dropdown" :key="option.value">
       <p x-text="option.text"
         @click="refreshDropdown()"
       ></p>
  </template>
</div>
```
and in my alpineJS function (which is not in reloaded html), i have something that look like that :

```
init()
{
    this.elDropdown = document.getElementById('dropdown');

    for(var index=0; index < this.elDropdown.options.length; index++)
    {
        this.dropdown.push(this.elDropdown.options[index]);
    }

    // reTrigger after each refreshDropdown()
    window.livewire.on('DropdownLivewireBooted', () => {

      for(var index=0; index < this.elDropdown.options.length; index++)
      {
          this.dropdown.push(this.elDropdown.options[index]);
      }
    });
},

refreshDropdown()
{
    // clean the current dropdown list
    this.dropdown = [];

    // reload the new dropdown list
    window.livewire.emitTo('dropdown-livewire', 'refreshDropdown');
},
```

So, what it does on each refreshDropdown() call, is :
    1.   reseting the dropdown property to an empty array
    2.   ask livewire (emitTo) (ajax start) to rebuild a new list(html: datalist & template tags)
    3.   then rebuild the dropdown property on Livewire DropdownLivewireBooted Event (ajax over)

But for some unknow reason, on certain subsequent request, i get a error : 
```
Alpine Expression Error : prevKeys is undefined
Expression: "dropdown"

<template x-for="(option,index) in dropdown" :key="option.value">
    _x_attributeCleanups: Object {  }
    _x_effects: Set []
    _x_keyExpression: "option.value"
    _x_runEffects: function _x_runEffects()
    accessKey: ""
    ...
```
And this happen even when the template have been successfully rendered ... but then, the javascript can't work anymore on subsequent calls because of that error.

Any hint on what could go wrong ???
i don't understand because my dropdown list is good ¯\\(°_o)/¯
what is that `prevKeys` ???

Thanks,

**Comment by SimoTod** [comment]
Stick a wire:ignore on the template tag or even on the parent div so Livewire won't try to replace that element. The prevKeys is a property that Alpine sets on the template element when it evaluates the request the first time so it gets lost if Livewire decides to replace that element for any reason.

**Comment by Ronan-Lenor** [comment]
i couldn't succeed either with `wire:ignore` or refactoring everything to end with `<template x-for="(item, key) in $wire.dropdown" :key="key">`. it always end, in some circumstance, with a huge mess (duplicate or old entries) in the generated new list

So, is it impossible to make a template x-for list been correctly dynamic?
is there not a way to make that template x-for completly forget its previous values and reinitialise it?

**Reply by Ronan-Lenor** [reply]
Well, i didn't wanted to do modification directly inside the way of how AlpineJS work, like suggested here : [https://github.com/alpinejs/alpine/discussions/163](https://github.com/alpinejs/alpine/discussions/1635)

So, i ended by targeting a `div` in javascript and update its content throught `innerHTML` with a `string` containing my `<template x-for="(item, index) in $wire.dropdown" ... `.

Then now, each time i update the list, the template x-for tag don't mess with its previous records.

it look like something like this:
```
init()
{
    // call a method in livewire to build the list
    this.$wire.buildDropdown();

    // from Livewire dehydrate method ( https://laravel-livewire.com/docs/2.x/lifecycle-hooks ) use : $this->emit('DropdownLivewireDehydrate');
    window.livewire.on('DropdownLivewireDehydrate', () => {
        document.getElementById('aDivTag').innerHTML = `
            <template x-for="(text, key) in $wire.dropdown">
                <p x-text="text" class="p-2"></p>
            </template>
        `;
    });
},
```
